src/Dfc.CourseDirectory.WebV2/Features/Venues/AddVenue/AddVenueController.cs

```diff
@@ -0,0 +1,193 @@
+        {
+            _journeyInstance.Delete();
+
+            if (_journeyInstance.Properties.TryGetValue(ReturnUrlJourneyInstancePropertyKey, out var returnUrlObj) &&
+                returnUrlObj is string returnUrl &&
+                Url.IsLocalUrl(returnUrl))
+            {
+                return Redirect(returnUrl);
+            }
+
+            return RedirectToAction(nameof(ViewVenues.ViewVenuesController.ViewVenues), "ViewVenues")
+                .WithProviderContext(_providerContextProvider.GetProviderContext());
+        }
+
+        [HttpGet("postcode-search")]
+        [RequireJourneyInstance]
+        public async Task<IActionResult> PostcodeSearch(PostcodeSearch.SearchCommand command) =>
+            await _mediator.SendAndMapResponse(
+                command,
+                result => result.Match(
+                    errors => this.ViewFromErrors(errors),
+                    vm => View("PostcodeSearchResults", vm)));
+
+        [HttpPost("select-postcode")]
+        [RequireJourneyInstance]
+        public async Task<IActionResult> SelectPostcode(PostcodeSearch.SelectCommand command, [FromQuery] string postcode)
+        {
+            command.Postcode = postcode;
+
+            return await _mediator.SendAndMapResponse(
+                command,
+                result => result.Match<IActionResult>(
+                    errors => this.ViewFromErrors("PostcodeSearchResults", errors),
+                    success => RedirectToAction(nameof(Details))
+                        .WithProviderContext(_providerContextProvider.GetProviderContext())
+                        .WithJourneyInstanceUniqueKey(_journeyInstance)));
+        }
+
+        [HttpGet("address")]
+        [RequireJourneyInstance]
+        public async Task<IActionResult> Address([FromQuery] bool? fromPublishPage) =>
+            await _mediator.SendAndMapResponse(
+                new Address.Query(),
+                vm => View(vm).WithViewData("FromPublishPage", fromPublishPage));
+
+        [HttpPost("address")]
+        [RequireJourneyInstance]
+        public async Task<IActionResult> Address(Address.Command command, [FromQuery] bool? fromPublishPage) =>
+            await _mediator.SendAndMapResponse(
+                command,
+                result => result.Match<IActionResult>(
+                    errors => this.ViewFromErrors(errors),
+                    success => RedirectToAction(fromPublishPage == true ? nameof(CheckAndPublish) : nameof(Details))
+                        .WithProviderContext(_providerContextProvider.GetProviderContext())
+                        .WithJourneyInstanceUniqueKey(_journeyInstance)));
+
+        [HttpGet("details")]
+        [RequireJourneyInstance]
+        public async Task<IActionResult> Details([FromQuery] bool? fromPublishPage) =>
+            await _mediator.SendAndMapResponse(
+                new Details.Query(),
+                vm => View(vm).WithViewData("FromPublishPage", fromPublishPage));
```

Comment:
Should we consider adding this value to the view model as opposed to using the ViewBag?

Reply:
I considered that but I'm not crazy about having the view model properties assigned in multiple places (action vs MediatR handler) and I don't want to thread through the property on the `Query` just to do it in the MediatR handler either.